src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/DistributedCacheTagHelperService.cs

```diff
@@ -74,19 +93,59 @@ HtmlEncoder HtmlEncoder
 
                             value = await _formatter.SerializeAsync(formattingContext);
 
-                            await _storage.SetAsync(key, value, options);
+                            
+                            using (var buffer = new MemoryStream())
+                            {
+                                var keyLength = BitConverter.GetBytes(serializedKey.Length);
+
+                                buffer.Write(keyLength, 0, keyLength.Length);
+                                buffer.Write(serializedKey, 0, serializedKey.Length);
+                                buffer.Write(value, 0, value.Length);
+
+                                await _storage.SetAsync(storageKey, buffer.ToArray(), options);
+                            }
 
                             content = formattingContext.Html;
                         }
                         else
                         {
-                            content = await _formatter.DeserializeAsync(value);
+                            // Extract the length of the serialized key
+                            byte[] contentBuffer = null;
+                            using (var buffer = new MemoryStream(value))
+                            {
+                                var keyLengthBuffer = new byte[sizeof(int)];
```

Comment:
I would make a constant for sizeof(int) and use that same constant on line 101 in place for keyLength.Length, if i'm not mistaken they represent the same concept and the constant makes it clearer